File: CSharp/Eta/TestTools/QATools/wlconsumer-MutiThreaded-001/ConsumerRequestThread.cs

//APIQA
namespace LSEG.Eta.ValueAdd.WatchlistConsumer;

using LSEG.Eta.Codec;
using LSEG.Eta.Common;
using LSEG.Eta.Example.Common;
using LSEG.Eta.Rdm;
using LSEG.Eta.Transports;
using LSEG.Eta.ValueAdd.Rdm;
using LSEG.Eta.ValueAdd.Reactor;
using System;
using System.IO;
using System.Net;
using System.Net.Sockets;
using static LSEG.Eta.Rdm.Dictionary;
using static Rdm.LoginMsgType;

/**
 * <p>
 * This interface is responsible for sending item requests to provider.
 * </p>
 */
class ConsumerRequestThread
{
    private Eta.Codec.Buffer? m_Payload;
    private ItemRequest? m_ItemReuest = null;
    private ChannelInfo? m_ChnlInfo;
    private ConsumerCallbackThread? m_ConsumerCallbackThread;

    private readonly ItemRequest m_ItemRequest = new();
    private readonly EncodeIterator m_EncIter = new();
    private readonly List<int> m_ViewFieldList = new();
    private readonly ReactorSubmitOptions m_SubmitOptions = new();

    public List<WatchlistConsumerConfig.ItemInfo>? m_ItemInfoList;
    public ConsumerRequestThread(ChannelInfo channelInfo, ConsumerCallbackThread consumer)
    {
        m_ItemReuest = CreateItemRequest();
        m_ChnlInfo = channelInfo;
        m_ConsumerCallbackThread = consumer;
        m_ItemInfoList = consumer!.m_WatchlistConsumerConfig!.ItemList!;
    }

    protected ItemRequest CreateItemRequest()
    {
        return new ItemRequest();
    }
    
    /* Runs the Value Add consumer application. */
    public void Run()		
	{		
		while (!m_ConsumerCallbackThread!.m_ShutDown)
		{

            //send  item requests
            if (m_ConsumerCallbackThread.IsRequestedServiceUp(m_ChnlInfo) && m_ConsumerCallbackThread.IsDictionaryReady()
                                   && !m_ConsumerCallbackThread.m_ItemsRequested)
            { 
                foreach (WatchlistConsumerConfig.ItemInfo item in m_ItemInfoList!)
    		    {
                    try
                    {
                        Thread.Sleep(10);
                        RequestItems(item);
                    }
                    catch (Exception e)
                    {
                        Console.WriteLine(e.ToString());
                    }
    		    }
                m_ConsumerCallbackThread.m_ItemsRequested = true;
            }
	        try
			{
				Thread.Sleep(10);
			} catch (Exception e)
			{
                Console.WriteLine(e.ToString());
            }
		}
		
		 // Handle run-time
        if (m_ConsumerCallbackThread.m_ShutDown)
        {
            Console.WriteLine("ConsumerRequestThread  " + this.GetHashCode() + " closing now...");
        }
	}

    private CodecReturnCode RequestItems(WatchlistConsumerConfig.ItemInfo? itemInfo)
    {
        CodecReturnCode ret = CodecReturnCode.SUCCESS;

        m_ItemRequest.Clear();

        if (m_ConsumerCallbackThread!.m_WatchlistConsumerConfig!.EnableSnapshot)
            m_ItemRequest.HasStreaming = true;

        m_ItemRequest.AddItem(itemInfo!.Name!);

        var domainType = itemInfo.Domain;

        m_ItemRequest.DomainType = domainType;
        m_ItemRequest.StreamId = itemInfo.StreamId;

        if (itemInfo.IsPrivateStream)
            m_ItemRequest.HasPrivateStream = true;

        if (domainType == (int)DomainType.SYMBOL_LIST && itemInfo.SymbolListData)
        {
            m_ItemRequest.IsSymbolListData = true;
            m_Payload = new();
            m_Payload.Data(new ByteBuffer(1024));

            m_EncIter.Clear();
            m_EncIter.SetBufferAndRWFVersion(m_Payload, m_ChnlInfo!.ReactorChannel!.MajorVersion, m_ChnlInfo!.ReactorChannel!.MinorVersion);

            ret = m_ItemRequest.Encode(m_EncIter);

            if (ret < CodecReturnCode.SUCCESS)
            {
                Console.WriteLine("RequestItem.Encode() failed");
                return ret;
            }

            m_ItemRequest.RequestMsg.EncodedDataBody = m_Payload;
        }
        else if (domainType == (int)DomainType.MARKET_PRICE && m_ConsumerCallbackThread!.m_WatchlistConsumerConfig!.EnableView)
        {
            m_Payload = new();// move it to the top to share
            m_Payload.Data(new ByteBuffer(1024));

            m_EncIter.Clear();
            m_EncIter.SetBufferAndRWFVersion(m_Payload, m_ChnlInfo!.ReactorChannel!.MajorVersion, m_ChnlInfo!.ReactorChannel!.MinorVersion);

            m_ItemRequest.HasView = true;
            m_ViewFieldList.Add(22);
            m_ViewFieldList.Add(25);
            m_ViewFieldList.Add(30);
            m_ViewFieldList.Add(31);
            m_ViewFieldList.Add(1025);
            m_ItemRequest.ViewFields = m_ViewFieldList;
            ret = m_ItemRequest.Encode(m_EncIter);

            if (ret < CodecReturnCode.SUCCESS)
            {
                Console.WriteLine("RequestItem.Encode() failed");
                return ret;
            }
            m_ItemRequest.RequestMsg.EncodedDataBody = m_Payload;
        }
        else
        {
            m_ItemRequest.Encode();
        }

        m_SubmitOptions.Clear();
        if (m_ConsumerCallbackThread!.m_WatchlistConsumerConfig!.ServiceName != null)
        {
            m_SubmitOptions.ServiceName = m_ConsumerCallbackThread!.m_WatchlistConsumerConfig.ServiceName;
        }

        var subRet = m_ChnlInfo!.ReactorChannel!.Submit(m_ItemRequest.RequestMsg, m_SubmitOptions, out var errorInfo);
        if (subRet < ReactorReturnCode.SUCCESS)
        {
            Console.WriteLine("\nReactorChannel.Submit() failed: " + ret + "(" + errorInfo?.Error.Text + ")\n");
            Environment.Exit((int)ReactorReturnCode.FAILURE);
        }

        return CodecReturnCode.SUCCESS;
    }
}

File: Java/Eta/TestTools/QATools/wlconsumer-MutiThreaded-001/ConsumerRequestThread.java
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.
 *|                See the project's LICENSE.md for details.
 *|           Copyright (C) 2025 LSEG. All rights reserved.
 *|-----------------------------------------------------------------------------
 */

package com.refinitiv.eta.valueadd.examples.watchlistconsumer;

import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.List;

import com.refinitiv.eta.codec.Buffer;
import com.refinitiv.eta.codec.CodecFactory;
import com.refinitiv.eta.codec.CodecReturnCodes;
import com.refinitiv.eta.codec.EncodeIterator;
import com.refinitiv.eta.rdm.DomainTypes;
import com.refinitiv.eta.valueadd.examples.watchlistconsumer.WatchlistConsumerConfig.ItemInfo;
import com.refinitiv.eta.valueadd.reactor.ReactorChannel;
import com.refinitiv.eta.valueadd.reactor.ReactorErrorInfo;
import com.refinitiv.eta.valueadd.reactor.ReactorFactory;
import com.refinitiv.eta.valueadd.reactor.ReactorReturnCodes;
import com.refinitiv.eta.valueadd.reactor.ReactorSubmitOptions;

/**
 * <p>
 * This interface is responsible for sending item requests to provider.
 * </p>
 */
public class ConsumerRequestThread implements Runnable
{
    private ReactorErrorInfo errorInfo = ReactorFactory.createReactorErrorInfo();
    private EncodeIterator encIter = CodecFactory.createEncodeIterator();
    private final List<Integer> viewFieldList;
    private ItemRequest itemRequest;
    private ReactorSubmitOptions submitOptions = ReactorFactory.createReactorSubmitOptions();
	boolean _finalStatusEvent;
	ChannelInfo chnlInfo;
	ConsumerCallbackThread consumerCallbackThread;
	Buffer payload;
	ArrayList<ItemInfo> itemList = new ArrayList<ItemInfo>();
	 
    public ConsumerRequestThread(ChannelInfo channelInfo, ConsumerCallbackThread consumer)
    {
    	 itemRequest = createItemRequest();
    	 viewFieldList = new ArrayList<Integer>();
        chnlInfo = channelInfo;
        consumerCallbackThread = consumer;
    }

    protected ItemRequest createItemRequest()
    {
        return new ItemRequest();
    }
    
    /* Runs the Value Add consumer application. */
    public void run()		
	{		
		while (!consumerCallbackThread.shutDown)
		{
	        //send  item requests
			if (consumerCallbackThread.isRequestedServiceUp(chnlInfo) && consumerCallbackThread.isDictionaryReady()
					&& !consumerCallbackThread.itemsRequested)
			{
    			for (ItemInfo itemInfo : itemList)
    			{
    				try 
    				{
    					Thread.sleep(10);
    					 requestItems(itemInfo);
    				}
    				catch (InterruptedException e)
    				{
    					e.printStackTrace();
    				}
    			}
    			consumerCallbackThread.itemsRequested = true;
    		}
    			
			if (consumerCallbackThread.isRequestedTunnelStreamServiceUp(chnlInfo)  && !consumerCallbackThread.itemsRequested)
			{	
					if (chnlInfo.connectionArg.tunnel()) 
					{
						if (!consumerCallbackThread.tunnelStreamHandler.isServiceFound())
						{
							System.out.println(" Directory response does not contain service name for tunnel streams: \n " 
									+ chnlInfo.connectionArg.tsService());
						}
						else if (!consumerCallbackThread.tunnelStreamHandler.isServiceSupported())
						{
							System.out.println(" Service in use for tunnel streams does not support them: \n"						 
									+ chnlInfo.connectionArg.tsService());
						}
			            else if (consumerCallbackThread.isRequestedTunnelStreamServiceUp(chnlInfo))
			            {
			                if (consumerCallbackThread.tunnelStreamHandler.openStream(chnlInfo, errorInfo) != ReactorReturnCodes.SUCCESS)
			                {
			                    if (chnlInfo.reactorChannel.state() != ReactorChannel.State.CLOSED &&
			                            chnlInfo.reactorChannel.state() != ReactorChannel.State.DOWN_RECONNECTING)
			                    {
			                    	consumerCallbackThread.shutDown(true);
			                        System.exit(ReactorReturnCodes.FAILURE);
			                    }
			                }
			            }
					}			
			}
			
	        try
			{
				Thread.sleep(10);
			} catch (InterruptedException e)
			{
				e.printStackTrace();
			}
		}
		
		 // Handle run-time
        if (consumerCallbackThread.shutDown)
        {
        	System.out.println("ConsumerRequestThread  " + this.hashCode() + " closing now...");
        }
	}
    
	int requestItems(ItemInfo itemInfo)
	{
	    int ret = CodecReturnCodes.SUCCESS;
	    
		itemRequest.clear();	
		
		if (!consumerCallbackThread.watchlistConsumerConfig.enableSnapshot())
			itemRequest.applyStreaming();

		itemRequest.addItem(itemInfo.name);
		
		int domainType = itemInfo.domain();
		
		itemRequest.domainType(domainType);
		itemRequest.streamId(itemInfo.streamId);

		if (itemInfo.isPrivateStream())
			itemRequest.applyPrivateStream();
					
		if (domainType == DomainTypes.SYMBOL_LIST && itemInfo.symbolListData)
		{
			itemRequest.symbolListData(true);
		    payload = CodecFactory.createBuffer(); 
		    payload.data(ByteBuffer.allocate(1024));

			if (payload == null)
			{
				return CodecReturnCodes.FAILURE;
			}

			encIter.clear();
			encIter.setBufferAndRWFVersion(payload, chnlInfo.reactorChannel.majorVersion(), chnlInfo.reactorChannel.minorVersion());
			
			
			ret = itemRequest.encode(encIter);
	         
			if (ret < CodecReturnCodes.SUCCESS)
			{
				errorInfo.error().text("RequestItem.encode() failed");
				errorInfo.error().errorId(ret);
				return ret;
			}

			itemRequest.requestMsg.encodedDataBody(payload);
		}
		else if (domainType == DomainTypes.MARKET_PRICE && consumerCallbackThread.watchlistConsumerConfig.enableView())
		{			
		    payload = CodecFactory.createBuffer(); // move it to the top to share 
		    payload.data(ByteBuffer.allocate(1024));

			if (payload == null)
			{
				return CodecReturnCodes.FAILURE;
			}
			
			encIter.clear();
			encIter.setBufferAndRWFVersion(payload, chnlInfo.reactorChannel.majorVersion(), chnlInfo.reactorChannel.minorVersion());

			itemRequest.applyHasView();				
			viewFieldList.add(22);
			viewFieldList.add(25);
			viewFieldList.add(30);
			viewFieldList.add(31);
			viewFieldList.add(1025);
			itemRequest.viewFields(viewFieldList);				
			ret = itemRequest.encode(encIter);
											         
			if (ret < CodecReturnCodes.SUCCESS)
			{
				errorInfo.error().text("RequestItem.encode() failed");
				errorInfo.error().errorId(ret);
				return ret;
			}							
			itemRequest.requestMsg.encodedDataBody(payload);				
		}
		else
		{
		    itemRequest.encode();
		}
				
		submitOptions.clear();
		if (consumerCallbackThread.watchlistConsumerConfig.serviceName() != null)
		{
		    submitOptions.serviceName(consumerCallbackThread.watchlistConsumerConfig.serviceName());
		}

		ret = chnlInfo.reactorChannel.submit(itemRequest.requestMsg, submitOptions, errorInfo);
		if (ret < CodecReturnCodes.SUCCESS)
		{
			System.out.println("\nReactorChannel.submit() failed: " + ret + "(" + errorInfo.error().text() + ")\n");
			System.exit(ReactorReturnCodes.FAILURE);
		}
		
		return CodecReturnCodes.SUCCESS;
	}	
}
